### PR TITLE
ICU-21402 replace sd and rg by subdivisionAlias

### PR DESCRIPTION
**Pull request by FrankYFTang**
Only replace if the replacement value fit alphanum{3,8} for now.
replacement value length of 2 are pending on resolution of
https://unicode-org.atlassian.net/browse/CLDR-14312

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21402
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/locid.cpp [hunk 1]
```diff
@@ -1559,11 +1646,40 @@ AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
...
+        changed = 0;
+        Locale temp(locale);
+        LocalPointer<icu::StringEnumeration> iter(locale.createKeywords(status));
+        if (U_SUCCESS(status) && !iter.isNull()) {
+            const char* key;
+            while ((key = iter->next(nullptr, status)) != nullptr) {
+                status = U_ZERO_ERROR;
+                CharString value;
+                CharStringByteSink valueSink(&value);
+                locale.getKeywordValue(key, valueSink, status);
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp:1667]: I propose that you get the keyword value only if the key is one of the two where we might replace the value.

File: icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java [hunk 2]
```diff
@@ -1268,12 +1268,31 @@ public String replace() {
...
                     ((variants == null) ? "" : Utility.joinStrings("_", variants)));
-                if (extensions != null) {
-                    result += extensions;
+            if (extensions != null) {
+                boolean keywordChanged = false;
+                ULocale temp = new ULocale(result + extensions);
+                Iterator<String> keywords = temp.getKeywords();
+                while (keywords != null && keywords.hasNext()) {
+                    String key = keywords.next();
+                    String value = temp.getKeywordValue(key);
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java:1282]: As in C++, I propose fetching the value only if the key is one of the two relevant ones.

**Comment by FrankYFTang** [comment]
PTAL

File: icu4c/source/common/locid.cpp [hunk 3]
```diff
@@ -1559,11 +1646,40 @@ AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
...
+                    CharStringByteSink valueSink(&value);
+                    locale.getKeywordValue(key, valueSink, status);
+                    if (U_FAILURE(status)) {
+                        status = U_ZERO_ERROR;
+                        continue;
+                    }
+                    if (replaceSubdivision(value, status)) {
+                        changed++;
+                    }
+                    temp.setKeywordValue(key, value.data(), status);
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp:1676]: Sorry, found a problem: if U_FAILURE after this line, then the loop just continues. Worst case, you get to the second relevant key and clobber the status back to a zero-error.

Please bail out of the loop after this line if there is a failure code. I am not looking at enough context to see if you just need to "break" or if you should return immediately with something.

**Comment by FrankYFTang** [comment]
PTAL

File: icu4c/source/common/locid.cpp [hunk 4]
```diff
@@ -1559,11 +1646,43 @@ AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
...
+    }
     if (extensionsStr != nullptr) {
+        changed = 0;
+        Locale temp(locale);
+        LocalPointer<icu::StringEnumeration> iter(locale.createKeywords(status));
+        if (U_SUCCESS(status) && !iter.isNull()) {
+            const char* key;
+            while ((key = iter->next(nullptr, status)) != nullptr) {
+                if (uprv_strcmp("sd", key) == 0 || uprv_strcmp("rg", key) == 0) {
+                    status = U_ZERO_ERROR;
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp:1665]: Is this actually necessary? I would remove this line.

File: icu4c/source/common/locid.cpp [hunk 5]
```diff
@@ -1559,11 +1646,43 @@ AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
...
+                    if (U_FAILURE(status)) {
+                        status = U_ZERO_ERROR;
+                        continue;
+                    }
+                    if (replaceSubdivision(value, status)) {
+                        changed++;
+                    }
+                    temp.setKeywordValue(key, value.data(), status);
+                    if (U_FAILURE(status)) {
+                      return false;
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp:1678]:
```suggestion
                        return false;
```

**Review by markusicu — COMMENTED** [review]
1 or 2 minor fixes, then please squash

**Comment by FrankYFTang** [comment]
PTAL

**Review by markusicu — APPROVED** [review]
lgtm

**Pull request merged**